commons-impl/src/main/java/com/connexta/transformation/commons/api/impl/AbstractTransformationImpl.java

```diff
@@ -0,0 +1,297 @@
+  public void add(AbstractMetadataImpl... metadatas) {
+    add(Stream.of(metadatas));
+  }
+
+  /**
+   * Adds the given metadata to the set contained in this transformation.
+   *
+   * @param metadatas the stream of metadata to add to the contained set
+   */
+  public void add(Stream<? super AbstractMetadataImpl> metadatas) {
+    metadatas
+        .map(AbstractMetadataImpl.class::cast)
+        .forEach(m -> this.metadatas.put(m.getMetadataType(), m));
+  }
+
+  @Override
+  public int hashCode() {
+    return hashCode0();
+  }
+
+  @Override
+  public boolean equals(Object obj) {
+    return equals0(obj);
+  }
+
+  @Override
+  public String toString() {
+    return String.format(
+        "AbstractTransformationImpl[id=%s, requestInfo=%s, startTime=%s, metadatas=%s]",
+        getId(), requestInfo, startTime, metadatas);
+  }
+
+  @Override
+  protected TransformationPojo writeTo(TransformationPojo pojo) throws PersistenceException {
+    if (hasUnknowns()) { // cannot serialize if it contains unknowns
+      throw new InvalidFieldException("unknown transformation");
+    }
+    super.writeTo(pojo);
+    convertAndSetOrFailIfNull(
+        "requestInfo", this::getRequestInfo, RequestInfoImpl::toPojo, pojo::setRequestInfo);
+    setOrFailIfNull("startTime", this::getStartTime, pojo::setStartTime);
+    return pojo.setVersion(TransformationPojo.CURRENT_VERSION)
+        .setMetadatas(
+            unwrap(() -> metadatas.values().stream().map(wrap(AbstractMetadataImpl::toPojo))));
+  }
+
+  @Override
+  protected final void readFrom(TransformationPojo pojo) throws PersistenceException {
+    super.readFrom(pojo);
+    if (pojo.getVersion() < TransformationPojo.MINIMUM_VERSION) {
+      throw new UnsupportedVersionException(
+          "unsupported "
+              + AbstractTransformationImpl.PERSISTABLE_TYPE
+              + " version: "
+              + pojo.getVersion()
+              + " for object: "
+              + getId());
+    } // do support pojo.getVersion() > CURRENT_VERSION for forward compatibility
+    this.hasUnknowns = pojo instanceof UnknownPojo; // reset the unknown flag
+    readFromCurrentOrFutureVersion(pojo);
+  }
+
+  /**
+   * Converts the specified metadata pojo into a metadata object.
+   *
+   * @param pojo the pojo to be converted
+   * @return the corresponding metadata object
+   * @throws InvalidFieldException if an error occurs while trying to deserialize the pojo object
+   * @throws PersistenceException if an error occurs while deserializing the pojo object
+   */
+  protected abstract AbstractMetadataImpl fromPojo(MetadataPojo pojo) throws PersistenceException;
+
+  @VisibleForTesting
+  void setRequestInfo(RequestInfoImpl requestInfo) {
+    this.requestInfo = requestInfo;
+  }
+
+  @VisibleForTesting
+  void setStartTime(Instant startTime) {
+    this.startTime = startTime;
+  }
+
+  @VisibleForTesting
+  int hashCode0() {
+    return Objects.hash(super.hashCode(), startTime, requestInfo, metadatas);
```

Comment:
❓ Can we not just put this implementation in the actual `hashCode()` method and spy this object for testing?